crates/wasmparser/src/collections/map.rs

```diff
@@ -0,0 +1,468 @@
+    Occupied(OccupiedEntry<'a, K, V>),
+    /// A vacant entry.
+    Vacant(VacantEntry<'a, K, V>),
+}
+
+impl<'a, K, V> Entry<'a, K, V>
+where
+    K: Ord,
+{
+    /// Returns a reference to this entry's key.
+    pub fn key(&self) -> &K {
+        match *self {
+            Self::Occupied(ref entry) => entry.key(),
+            Self::Vacant(ref entry) => entry.key(),
+        }
+    }
+}
+
+/// A view into an occupied entry in a [`Map`].
+///
+/// It is part of the [`Entry`] enum.
+#[derive(Debug)]
+pub struct OccupiedEntry<'a, K: Ord, V> {
+    inner: detail::OccupiedEntryImpl<'a, K, V>,
+}
+
+impl<'a, K: 'a, V: 'a> OccupiedEntry<'a, K, V>
+where
+    K: Ord,
+{
+    /// Gets a reference to the key in the entry.
+    pub fn key(&self) -> &K {
+        self.inner.key()
+    }
+
+    /// Gets a reference to the value in the entry.
+    pub fn get(&self) -> &V {
+        self.inner.get()
+    }
+
+    /// Gets a mutable reference to the value in the entry.
+    pub fn get_mut(&mut self) -> &mut V {
+        self.inner.get_mut()
+    }
+
+    /// Sets the value of the entry with the [`OccupiedEntry`]'s key, and returns the entry's old value.
+    pub fn insert(&mut self, value: V) -> V {
+        self.inner.insert(value)
+    }
+
+    /// Converts the [`OccupiedEntry`] into a mutable reference to the value in the entry
+    /// with a lifetime bound to the map itself.
+    pub fn into_mut(self) -> &'a mut V {
+        self.inner.into_mut()
+    }
+}
+
+/// A view into a vacant entry in a [`Map`].
+///
+/// It is part of the [`Entry`] enum.
+#[derive(Debug)]
+pub struct VacantEntry<'a, K: Ord, V> {
+    inner: detail::VacantEntryImpl<'a, K, V>,
+}
+
+impl<'a, K: 'a, V: 'a> VacantEntry<'a, K, V>
+where
+    K: Ord,
+{
+    /// Gets a reference to the key in the entry.
+    pub fn key(&self) -> &K {
+        self.inner.key()
+    }
+
+    /// Take ownership of the key.
+    pub fn into_key(self) -> K {
+        self.inner.into_key()
+    }
+
+    /// Sets the value of the entry with the [`VacantEntry`]'s key, and returns a mutable reference to it.
+    pub fn insert(self, value: V) -> &'a mut V
+    where
+        K: Hash,
+    {
+        self.inner.insert(value)
+    }
+}
+
+impl<K, V> FromIterator<(K, V)> for Map<K, V>
+where
+    K: Hash + Eq + Ord,
+{
+    fn from_iter<I>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = (K, V)>,
+    {
+        Self {
+            inner: <detail::MapImpl<K, V>>::from_iter(iter),
+        }
+    }
+}
+
+impl<'a, K, V> IntoIterator for &'a Map<K, V> {
+    type Item = (&'a K, &'a V);
+    type IntoIter = Iter<'a, K, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+/// An iterator over the items of a [`Map`].
+#[derive(Debug, Clone)]
+pub struct Iter<'a, K, V> {
+    inner: detail::IterImpl<'a, K, V>,
+}
+
+impl<'a, K, V> Iterator for Iter<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
```

Comment:
Mind adding `size_hint` overrides which forward to the underlying iterator as well to this and other impls?

Reply:
Yes, I also did this in my Wasmi implementation already.

Reply:
done